New block: Add a Breadcrumbs block

andrewserong opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Related to #21943

This PR proposes adding in a new Breadcrumbs block, to render navigation links of the hierarchy of a page. This is a server rendered block, which would typically be used in a page template, most likely before or after the Post Title.

## Features include

* Markup follows this example: https://www.w3.org/WAI/ARIA/apg/patterns/breadcrumb/examples/breadcrumb/
* Support for pages (hierarchical post types) _or_
* Support for displaying a single hierarchical category as breadcrumbs for posts
* Set a custom separator character (click on the separator in the editor canvas to edit)
* Show leading separator (so that the breadcrumb can be rendered like a back button)
* Show current page / post title
* Show site title
* Custom override text to override the site title (click on the site title in the editor canvas to edit)
<!-- * Schema.org markup is rendered on the server following this [example](https://developers.google.com/search/docs/data-types/breadcrumb#microdata) -->

Some of these features are implemented in a subtle way so as to not be intrusive. To override the site title, once it's displayed as the first breadcrumb, a user can click into the site title and manually update it. Likewise, the separator character defaults to `/` but can be edited by clicking on it and replacing it with another character.

<!--

## Design / feedback sought

* [ ] What is the Desired behaviour / layout on mobile / narrow viewports?
* [ ] Feedback on all of the placeholder text both within the block in the editor canvas (placeholder names for the breadcrumbs), and in the text in the sidebar
* [ ] Do we need a custom design for the block's icon, since I don't think we have any existing icons that are really suitable (unless anyone can think of one)?
* [ ] Discoverability, order, and behaviour of the block and its settings

## Still to do

* [ ] Add tests
* [ ] Should separator color be a separate control from text color? (Say if we want the current post title to not be a link)
* [ ] Figure out how the breadcrumbs should collapse on mobile / wrap when it's quite long
* [ ] Fine-tune the wording on each of the controls / strings introduced
* [ ] Add a limit to how far we iterate over parent ids in the editor view (protect against circular parent relationships?)
* [ ] Do we need a different icon for this block? (For the block inserter)
* [ ] Set placeholder for nesting level of `0` to a more meaningful word, e.g. `All`
* [ ] Add logic / UI messaging to deal with the potential case where the site title is empty ([comment](https://github.com/WordPress/gutenberg/pull/32500#issuecomment-921726373))

## Beyond MVP

* [ ] Can we get this to support custom taxonomies? E.g. so that plugins dealing with ecommerce or custom post types that use their own taxonomy such as a Product Categories can use the core block and have it either "just work" or be simple to hook up (e.g. a hierarchical taxonomy selector)
* [ ] Can we support custom icons or images in the custom separator field?

-->

## Testing instructions

<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

1. Use a block-based theme such as TT3
2. Open up the site editor and edit a template part that will be used for pages (or any hierarchical post type)
3. Insert the Breadcrumbs block to the template part so that it's rendered when you view a page on the front end of the site
4. Note that in the site editor, the block will display placeholder names for each of the breadcrumbs
5. Create a couple of pages, and assign parents to these pages so that you have nested pages on your site
6. View the page on the front end of the site, and you should see the Breadcrumbs rendered in the appropriate place in the template

Repeat the above with a normal post, and use hierarchical post types.

If you wish to test the custom block spacing control, be sure to use an FSE theme, and set the `settings.spacing.blockGap` value to `true` in your `theme.json` file (for more information on this see the testing instructions in a blockGap PR like #34630)

## Screenshots <!-- if applicable -->

<!--

This is a bit of an extreme example, but here is the front end of a site when the block is used as part of a page template, before the post content:

![image](https://user-images.githubusercontent.com/14988353/137261948-7daf7ae2-f31c-4035-ba1c-bee867e3702f.png)

This block should support a variety of styles / patterns for breadcrumbs. Some examples include a typical use case where each item is separated by a slash. But we can also limit the nesting level to one, and prefix it with a back arrow to effectively create a back button/link:

| A few different uses | Settings in the editor |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/121306117-d051a180-c941-11eb-81de-3fd284392d78.png) | ![image](https://user-images.githubusercontent.com/14988353/137262229-fc2d763c-a9b1-4c1f-af0a-3abbd6469758.png) |

## Inspiration / the end goal

The objective with this new block is to support a variety of different designs for breadcrumbs, so that the block is flexible for use in many different kinds of block patterns. To provide an example for the kind of flexibility we'd like to achieve with this block, here are some screenshots of breadcrumbs in use on a few different museum websites, as inspiration:

| [British Museum](https://www.britishmuseum.org/exhibitions/hokusai-great-picture-book-everything) | [The Frick Collection](https://www.frick.org/exhibitions/bertoldo) |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/137058899-b8373a16-3860-4564-9277-90958dd7953e.png) Override site title, and include current page title | ![image](https://user-images.githubusercontent.com/14988353/137059062-e460df4c-eb6d-401d-96e9-c8b7e33213bf.png) Override site title, don't include current page title |

| [ACMI](https://www.acmi.net.au/whats-on/story-of-the-moving-image-exhibition/) | [The Met](https://www.metmuseum.org/toah/hd/adtx/hd_adtx.htm) |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/137059201-e94cf40b-5c89-41f2-a99d-5e5bda637963.png) Use leading separator, only show 1 nesting level, don't included current page title | ![image](https://user-images.githubusercontent.com/14988353/137059627-524437cd-7a01-47ed-aff8-666c81bb9d14.png) Nesting level set to 2, custom separator character, don't show current page title, use center alignment |

-->

![image](https://github.com/WordPress/gutenberg/assets/14988353/c1d85b8f-2692-4de4-ae75-b7c77683cffd)

https://github.com/WordPress/gutenberg/assets/14988353/92e15e00-ab79-4582-8886-a75f852af496

andrewserong committed:
Add Breadcrumbs block initial implementation

andrewserong committed:
Add editor style support and options for leading separator and to show current page title

andrewserong committed:
Hook up editor view to real page titles and parents, fall back to placeholder titles

andrewserong committed:
Implement nesting level support within editor UI and in server rendering

andrewserong commented:
Update: I'm still keen to continue developing this block (I switched over to working on some other projects these past few weeks, but hope to keep chipping away at this in my spare time). If anyone has a chance to take a look, I'm keen to get feedback on the approach, and input into some design considerations:

* Does the approach of this server-rendered block seem okay?
* At the moment it just deals with hierarchical post types (e.g. pages). It'd be nice for it to also optionally handle hierarchical categories, so that we could use it for a template part that covers normal blog posts.
* Any ideas on how best to handle / break down the UI for mobile?
* What about placeholder text in the UI? At the moment if we can't grab real hierarchical post titles in the editor, I'm using some placeholder names so that editing a template part is visually as close to the real thing as we can get. Any ideas for better placeholder names?
* Should we also offer a selection of icons for the separator, so that we're not limited by characters available in the current font? In the sidebar, I'm imagining a few different icons that the user can click, and an "Other:" option where they can enter a custom character?
* What color controls do we need at the outset? (E.g. if the current page title is not a link, then we might want to be able to have a different color for the separator versus text color)

Many of these ideas might not be required for an MVP for this block, so I'm happy to hear any ideas!

andrewserong committed:
Fix linting issues in PHP file

andrewserong committed:
Add block test fixtures

andrewserong committed:
Remove editor store dependency

andrewserong commented:
I've given this a rebase and fixed up some linting issues, so this should be good for folks to take for a spin if you'd like to test it out. There's still a bunch of To-do list items (and likely many To-do items that I haven't thought of), so please do let me know if you have ideas on how you think this block should be implemented, or needed features.

andrewserong committed:
Rearrange markup so that separator comes after the anchor except for the leading separator

andrewserong committed:
Fix PHP linting issues

andrewserong committed:
Add schema.org markup using microdata

paaljoachim commented:
Hey @andrewserong 

Testing the Gutenberg PR build. This looks really really nice! Great work!

I made 3 nested pages: top page - 2 page - 3 page

![Screen Shot 2021-08-17 at 16 54 34](https://user-images.githubusercontent.com/5323259/129749478-14c6bfbe-1f53-42e2-9dd5-078a924d7ac6.jpg)

Panels:
Typography
Color
Spacing
Breadcrumbs settings (Should this be called something else? As all the sidebar when Breadcrumbs is selected is Breadcrumbs settings. Not sure what to call it at the moment.)

Nesting level. I am wondering if it should start without the value 0 present (if possible) or a dash - (no value). As I initially read it as Nesting level 0 (meaning no nesting levels). I then read the text: Set to 0 to show all nesting levels.

Nesting level does not show the home page.

![Screen Shot 2021-08-17 at 17 12 06](https://user-images.githubusercontent.com/5323259/129752506-587f1612-8744-4564-aa01-adac505d80e5.jpg)

I would assume when selecting Nesting level 0 that the home page would also show.
I would also assume that I can rename the home page breadcrumb link if I want to or use a house icon.

Separator character. When removing / I notice the placeholder says e.g. /
Perhaps it should say: e.g. / - 

Perhaps something like this:
![Screen Shot 2021-08-17 at 17 07 13](https://user-images.githubusercontent.com/5323259/129751635-5dc86577-6b92-46a7-9185-33bcd4633bb2.jpg)

I closed the various panels such as Typography and Spacing. Deselected the Breadcrumbs block and reselected it again. I noticed that Typography and Spacing was open again. I assume there is save mechanism so that when I close/open panels it would remain that way even upon deselecting and reselecting the Breadcrumbs block.

Andrew this is really really nice work! Thank you for working on it!

Btw the block toolbar alignment controls work really well!


I found this interesting resource:
https://semantic-ui.com/collections/breadcrumb.html

andrewserong commented:
Thanks for the feedback @paaljoachim, great suggestions!

 > Breadcrumbs settings (Should this be called something else

Yes, I couldn't quite think of a good name for it, either, but even "Breadcrumbs" would be better to drop the superfluous word "settings".

> Nesting level does not show the home page.

Good catch, I like the idea of adding that in, and I think you're right, it would be good to handle that implicitly as part of the nesting levels. It might be worth also having a toggle for whether or not to show it? Either way, I like your idea of being able to override the site title / home breadcrumb 👍

Icons would be a cool addition, too, I'm curious how far we can get without introducing icons as a first pass, and perhaps treat icons as an enhancement in a follow-up.

Thanks again for the feedback! I'm mostly focusing on helping out with some of the design tooling at the moment, but will come back to this and chip away at the suggestions here, let me know if you think of any others 🙂

paaljoachim commented:
Hey Andrew.

Using the name Breadcrumbs instead of Breadcrumbs settings seems fine (for now anyway).

A toggle to show the home page title seems like a good idea. Then a way to override the site title / home breadcrumb.

> Icons would be a cool addition, too, I'm curious how far we can get without introducing icons as a first pass, and perhaps treat icons as an enhancement in a follow-up.

I agree. It sounds like a nice follow-up for a later point. 
I am also wondering when talking about icons if adding the icons block https://github.com/WordPress/gutenberg/issues/16484 would be helpful for making some icons available in the Breadcrumbs block. 

I will let you know of additional suggestions as they show up.

andrewserong committed:
Update panel body title and separator placeholder

andrewserong committed:
Add option to show site title and override its label in JavaScript

andrewserong committed:
Implement site title option in server-side rendering

andrewserong committed:
Move block fixtures to match current expected location

andrewserong committed:
Fix issue with extra separator being rendered when current page title is switch to show

andrewserong commented:
I had a bit of time at the end of the week to make a couple of additions to the PR (thanks again for the feedback, Paal!):

* Added "Show site title" option
* Added "Site title override" option that's only visible when "Show site title" is selected
* Got the above to render correctly in the editor and in the PHP server-rendering
* Renamed the panel heading to "Breadcrumbs"

Here's how it looks now in the editor:

![image](https://user-images.githubusercontent.com/14988353/130178273-8a2821a1-74e3-4914-9334-d2c915463af0.png)

I'm sure there'll be some tweaks needed, but it feels nice having the site title in there now. I'll be working on other things next week, but hope to pick this PR up again after that!

andrewserong committed:
Move test fixtures up a level

andrewserong committed:
Update test fixture with new default showSiteTitle attribute

andrewserong committed:
Fix linting issues

paaljoachim commented:
Looks good Andrew!

Retesting.
on - Show leading seperator
on - Show current page title
on - Show site title

![Screen Shot 2021-08-22 at 14 05 54](https://user-images.githubusercontent.com/5323259/130354467-f098d667-ff15-429c-ae88-771c18412ce4.jpg)

I changed the home page to Front page.

Even though underlines should be in place because of accessibility it would be nice to have an option to turn these off.
Here is an associated PR.
**Post Title Block: add typography formatting options**
https://github.com/WordPress/gutenberg/pull/31623

Here is an example from my dev site where I am recreating the main site:
https://new.easywebdesigntutorials.com/design-ui-ux-resources/

![Screen Shot 2021-08-22 at 14 08 48](https://user-images.githubusercontent.com/5323259/130354563-57e3ae8b-544e-43b0-938f-a5b107fccde9.jpg)

Here we can see none active links green while the active link contains the default black color. NB. It also uses all capital letters which I assume will be in the typography controls.
Should there be a way to add colors to active and none active links? Similar to the above example. 
Here is an associated PR.
**[WIP] Block Support: Custom secondary colors support**
https://github.com/WordPress/gutenberg/pull/33157

andrewserong commented:
Great suggestions, thanks for testing again, Paal! Yes, I think it'd be great to be able to control the display of underlines and custom secondary colors to support those sorts of designs, I imagine they'll be very common patterns. I'll dive in when I can circle back to this PR 🙂

andrewserong committed:
Opt in to the font weight, style, and text transform typography controls

andrewserong commented:
Just pushed a small update to opt the block in to the font weight, style, and text transform typography controls. This gives a bit more flexibility to the variety of designs the block can support, from the creative to the slightly obnoxious 😄. Here's what it currently looks like:

![image](https://user-images.githubusercontent.com/14988353/133725868-08bd23df-57df-475f-a3fb-2d784de03088.png)

For the underline, it'll be great to be able to control it. I see there's an open WIP PR in https://github.com/WordPress/gutenberg/pull/31768 to add in that typography control, too.

paaljoachim commented:
Awesome work Andrew! This is looking better and better! Thank you!

I am doing some testing.
At the moment it looks like the "Show site title" is not being seen.
Also align Center freezes the Change text Alignment to center.

Alignment: center. Text alignment does not work.
![Screenshot 2021-09-17 at 13 32 06](https://user-images.githubusercontent.com/5323259/133775818-d2069d56-e6e3-459f-aabd-da9bde32e119.jpg)

Show site title. Looks like it is not working, as it is not showing the Front page title I set. Also removing the custom Site title does not automatically add in the Home title.

![Screenshot 2021-09-17 at 13 32 35](https://user-images.githubusercontent.com/5323259/133775881-bf9aea60-d1e6-4745-afd1-fb3d55c02725.jpg)


O btw
I am not seeing the Border settings panel.
![Screenshot 2021-09-17 at 13 36 08](https://user-images.githubusercontent.com/5323259/133776283-ff35022e-1a65-45d7-a1cb-dd495aa188c7.jpg)

I am testing with downloading the special PR Gutenberg build.

andrewserong committed:
Update block.json alignment to only set wide and full width

andrewserong commented:
Thanks for the additional testing @paaljoachim, much appreciated!

> Also align Center freezes the Change text Alignment to center.

I've pushed an update to the `block.json` file to only opt the block in to the `wide` and `full` alignments, so that the left/center/right alignments are handled by the text alignment control instead. This should avoid the situation where you can get stuck using the block-level center alignment, and makes the block more consistent with similar blocks like Post Title.

> I am not seeing the Border settings panel.

I believe border controls are only displayed on themes that use a `theme.json` file and opt-in to that block support. TT1-Blocks uses the following in its `settings` key, which switches on the border control, and so exposes it on this block, too:

<img src="https://user-images.githubusercontent.com/14988353/133951548-0f453326-8807-4121-8152-2236ebcc2fc2.png" width="300" />

> Show site title. Looks like it is not working, as it is not showing the Front page title I set. Also removing the custom Site title does not automatically add in the Home title.

Very strange! I haven't been able to replicate this yet in my test environment, but I'll do some more digging. Can you please share the following:

* In your test site, what is the name that the site is currently set to? E.g. the Site Title under `/wp-admin/options-general.php`
* Which theme are you currently using for testing?

Thanks again, it's very helpful having someone test this out while I'm tinkering away at this PR! 🙇

andrewserong committed:
Add support for post categories

andrewserong committed:
Fix whitespace

andrewserong commented:
I've just pushed a change to explore adding in support for post categories, too. The logic now attempts to use the post/page hierarchy if available and then, if not available, it'll fall back to attempting to use the first _category_ attached to the post. Here's how it looks in TT1-blocks currently on a post with a hierarchical category added:

![image](https://user-images.githubusercontent.com/14988353/133962182-2f2f468c-9881-4f76-be33-6b25da3b2608.png)

In the above screenshot, the hierarchy goes:

1. Site title
2. Travel (parent category)
3. Australia (sub-category of "Travel" category)
4. Current post title

With the additional logic, some of the functions are now getting a little long, so before this PR comes out of Draft and when it's a little further along, I might spend some time cleaning / tightening up the code.

paaljoachim commented:
Hello Andrew

> I've pushed an update to the block.json file to only opt the block in to the wide and full alignments, so that the left/center/right alignments are handled by the text alignment control instead. This should avoid the situation where you can get stuck using the block-level center alignment, and makes the block more consistent with similar blocks like Post Title.

Example the Heading block has a wide and full alignment and can also align text left, center or right.

![Screenshot 2021-09-20 at 08 08 33](https://user-images.githubusercontent.com/5323259/133962412-3262fad3-90b7-413d-b51b-672a5f193734.jpg)

We should actually do the same for the Paragraph block as is done for the Heading block. Seeing the wide/full alignment and text alignment I thought that was a good idea as it would give better alignment options just like the Header block. One can e.g. have a wide block with a wide background color, a border with a color and border radius. Within this wide box one can then align the text. The text alignment should be spread out to more blocks, as it gives additional customization options for the user.

> I believe border controls are only displayed on themes that use a theme.json file and opt-in to that block support. TT1-Blocks uses the following in its settings key, which switches on the border control, and so exposes it on this block, too:

Example the Buttons block as a Border and Dimensions panels.
![Screenshot 2021-09-20 at 08 12 18](https://user-images.githubusercontent.com/5323259/133962616-4d3a846a-6639-4133-b444-d5d6eb86332e.jpg)

I am using the theme Twenty Twenty One.
Gutenberg plugin 11.5.0
WordPress 5.8.1

Site URL:
![Screenshot 2021-09-20 at 08 14 23](https://user-images.githubusercontent.com/5323259/133962719-9ceaff99-2cad-4bcf-a080-1c93c52054ce.jpg)

AHA! I see the Site Title is blank. Well that is a good explanation to why it is not working. (I will retest again at a later time to really see that not having a Site Title also breaks the site title feature in Breadcrumbs.)

andrewserong commented:
Thanks for the extra info! Glad we got to the bottom of the site title issue, perhaps we can add some error handling in the Breadcrumbs UI for the site title to handle the case that the site title is empty 👍

andrewserong commented:
> Example the Heading block has a wide and full alignment and can also align text left, center or right.

Just to clarify, this PR now behaves like the Heading block. The block-based alignment allows for `wide` and `full`, and a separate text alignment control allows `left`, `center` and `right`. I think the buggy behaviour was due to the block-level alignment for left/center/right overriding the text alignment, so this PR now behaves more like the Heading block for consistency.

andrewserong commented:
@paaljoachim just a follow-up on the border controls. In Gutenberg, the base `theme.json` provides default settings that all themes can inherit from. For the Button block, it's set to display the radius control by default across all themes, unless the theme opts-out of it. It's set by default [on this line](https://github.com/andrewserong/gutenberg/blob/239babc2825bba33f535f5e1d0f809b4e5c20606/lib/theme.json#L230-L230). The current blocks where the border controls are displayed by default are "Button" (just the custom radius), and "Pull quote" (color, radius, style, and width). (This is why we can use these controls on the TwentyTwentyOne theme even though it lacks its own `theme.json` file)

Once the UI for the border controls stabilises, I believe the plan is to switch many more blocks over to displaying the UI controls by default (which would ultimately also include this block). This PR uses the same block supports mechanism to opt-in to the controls as the Button block does, so if and when this PR lands, we'll be able to go through a similar process and add a small follow-up PR to switch on the controls by default across all themes.

In the mean-time, themes can opt-in to experimental controls, so I typically use TT1-Blocks as my testing theme (and manually edit the `theme.json` as I go) as it's a good way to test out these features before they stabilise.

paaljoachim commented:
Thanks for the clarifications Andrew!

paaljoachim commented:
Hey Andrew. I am working on a client site that uses ThemeX and their page builder Cornerstone. I am right now adding in breadcrumbs. This is what it looks like there.

For inspiration.

Home Link -Text-
![Screenshot 2021-09-30 at 13 50 29](https://user-images.githubusercontent.com/5323259/135450131-a4ae38e2-a12a-470b-922e-1a915d3d0c0e.jpg)

Home Link -Icon-
![Screenshot 2021-09-30 at 13 50 37](https://user-images.githubusercontent.com/5323259/135450197-e6a54821-586b-4041-b785-422ae57a6df2.jpg)

andrewserong commented:
Fantastic, thanks for the screenshots, Paal! Great to see an example of how others manage a UI for selecting icons, that looks like a neat way to handle it. I like the detail that even if "Icon" is selected, the label is still editable, too (I assume for accessibility).

paaljoachim commented:
Hey Andrew @andrewserong 
From earlier testings I believe we are at a point where we can call this the first version of the Breadcrumbs block, or is there something missing that we should have included?
(Later on I will retest.)

andrewserong commented:
Thanks for the ping @paaljoachim. In terms of features, yes I agree, I think we've honed in on a pretty good feature set for v1 and can defer icon support to follow-ups later on.

There are a couple of things I'd like to tweak this week (adding gap support between each of the breadcrumbs and ensure there is a gap if no separator is used).

Once I've tidied it up a little, I think we'll then be able to switch the PR out of draft mode and ping some designers for feedback. The primary areas I think we'll need some design assistance are:

* Desired behaviour / layout on mobile / narrow viewports
* Feedback on placeholder text both within the block in the editor canvas, and in the text in the sidebar
* A custom design for the block's icon, since I don't think we have any existing icons that are really suitable (unless anyone can think of one)

I'm just wrapping up some work on bug fixes today and tomorrow, but then hope to pick this up for a little bit to clean things up and update the PR description ready for others to take a look. I'm not sure where this block sits in the priorities for 5.9, but it'll be good to get some more pairs of eyes on it and see what folks think.

karmatosed commented:
It would be great to get some design feedback on this so I've popped the label on. @shaunandrews awesomely flagged this in the design channel and I'll also dive in to give a review.

My biggest feedback so far is this seems to have quite a few features. I wonder what could be the distilled version and what could we lean into block styles even for? I am not absolutely convinced it needs everything.

paaljoachim commented:
I test this PR by using this approach: https://make.wordpress.org/design/2021/03/03/testing-a-gutenberg-pull-request-pr/

Going through the current features.

Default sidebar settings (I believe that these will be default sidebar settings for all blocks that contain text.):
Typography
Color
Dimensions

![Screenshot 2021-10-12 at 22 42 39](https://user-images.githubusercontent.com/5323259/137026244-78d87be2-70dc-4f20-acce-8580e29d118e.jpg)


Breadcrumbs sidebar settings:
Nesting level --> Is helpful to have so that we can define how many levels to nest. For instance if only want to nest 2 levels we can do so.

Separator character --> Is helpful as we can separate with / - or any other character. As the user can choose how to separate the breadcrumb levels.

Show leading separator ---> Totally ok. (Not sure of the purpose. I assume to create consistency with how each breadcrumb looks like.)

Show current page title --> Is helpful as one can decide to show the current page as part of the breadcrumbs path.

Show site title --> Is helpful so we can choose to show the home page site title. Or decide to call it another name.

![Screenshot 2021-10-12 at 22 42 59](https://user-images.githubusercontent.com/5323259/137026360-a94c314b-9fb6-43c7-b6f0-4edccd4c1929.jpg)

andrewserong commented:
Thanks for adding the Needs Design Feedback label @karmatosed!

> My biggest feedback so far is this seems to have quite a few features. I wonder what could be the distilled version and what could we lean into block styles even for? I am not absolutely convinced it needs everything.

Most of the features are a result of looking at examples of breadcrumbs out in the wild and seeing that there's quite a lot of variation in how they're displayed. Ultimately, I imagine users will often use the breadcrumbs block as a part of a block pattern that already sets a number of these parameters, so I think of these controls as more of an "Advanced settings" kind of area. Perhaps we should rename that panel to "Display settings" for consistency with the Navigation block?

For context, and to expand on some of Paal's points, here are some examples of breadcrumbs from museum websites, and a short description of the settings we could use in the Breadcrumbs block if we were attempting to recreate these in block patterns:

| [British Museum](https://www.britishmuseum.org/exhibitions/hokusai-great-picture-book-everything) | [The Frick Collection](https://www.frick.org/exhibitions/bertoldo) |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/137058899-b8373a16-3860-4564-9277-90958dd7953e.png) Override site title, and include current page title | ![image](https://user-images.githubusercontent.com/14988353/137059062-e460df4c-eb6d-401d-96e9-c8b7e33213bf.png) Override site title, don't include current page title |

| [ACMI](https://www.acmi.net.au/whats-on/story-of-the-moving-image-exhibition/) | [The Met](https://www.metmuseum.org/toah/hd/adtx/hd_adtx.htm) |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/137059201-e94cf40b-5c89-41f2-a99d-5e5bda637963.png) Use leading separator, only show 1 nesting level, don't included current page title | ![image](https://user-images.githubusercontent.com/14988353/137059627-524437cd-7a01-47ed-aff8-666c81bb9d14.png) Nesting level set to 2, custom separator character, don't show current page title, use center alignment |

andrewserong committed:
Move buildBreadcrumb util to a React component instead

andrewserong committed:
Switch to flex, add gap support, switch to justification instead of alignment

andrewserong committed:
Fix linting issue

andrewserong commented:
There is still more to do on this block (and I'm sure there'll be plenty of feedback), but I've switched this PR out of Draft now, as it's taking a more stable shape as a candidate for a v1 of this block.

I have added in block gap support for themes that enable it, switched the layout to use `flex`, and reworded the PR description so that there's more structure, and a list of the current features (and those we hope to implement either soon, or in a v2 / follow-up). I've also copied the examples of breadcrumbs in other sites to the description to provide better context for the goal we're aiming for in terms of the feature set.

karmatosed commented:
@andrewserong I tend to think of this a little differently:

> Most of the features are a result of looking at examples of breadcrumbs out in the wild and seeing that there's quite a lot of variation in how they're displayed. Ultimately, I imagine users will often use the breadcrumbs block as a part of a block pattern that already sets a number of these parameters, so I think of these controls as more of an "Advanced settings" kind of area. Perhaps we should rename that panel to "Display settings" for consistency with the Navigation block?

Why not just have styles?

I think right now in testing the block is very overwhelming and whilst I understand the review of other examples, this shouldn't be the 'one breadcrumb example that solves everything'. What is the minimum breadcrumb block that just works and is great?

andrewserong commented:
Thanks for taking a look @karmatosed!

> Why not just have styles?

The main reasons I'm leaning toward controls instead of styles are:

* An end user (not a developer / or theme designer) should be able to customise the display and separator character used without having to push any code
* For consistency with the goals of full site editing, a user shouldn't need to install a plugin or switch themes to be able to control the behaviour of the individual block
* Aside from the separator character used, the settings alter the semantic markup of the Breadcrumbs block, including the schema.org markup, so it's better suited to a settings / attributes change, than a CSS class
* Relying on styles instead of settings limits the flexibility of what folks can create for shareable block patterns

> I think right now in testing the block is very overwhelming and whilst I understand the review of other examples, this shouldn't be the 'one breadcrumb example that solves everything'.

That's a very good point, the sidebar UI feels pretty messy to look at to me, too! On the other hand, I think in terms of functionality, it's important that we build a core block with enough flexibility that it'll be useful for folks to use in blocks-based themes without needing to reach for a plugin.

If we can, rather than remove functionality, I'd be curious to see how we can have the same level of control, but communicate / expose those controls in a way that's more straightforward or easy to understand for a user. For example, if the nesting level control were to be reworked to be in the block toolbar instead of the sidebar, we could have the slider be closer to where it's affecting the display of the block, which would tidy up the sidebar. Or, if we want advanced controls to be hidden, we could move the least used controls to the Advanced panel in the sidebar where they're collapsed by default.

> What is the minimum breadcrumb block that just works and is great?

I struggled to find a good consistent default that is common across most sites that I researched in looking for examples of breadcrumbs. There's just so many variations! One way could be to come up with what we think is a sensible set of defaults that most folks will use, and then de-prioritise the display settings in the sidebar by having them collapsed by default.

We have a few precedents of site editing blocks that allow fine-tuned control (e.g. Query Loop, Navigation, Post Author), so I'm keen for us to keep that control if we can, but to carefully design how we expose that level of control.

andrewserong commented:
> One way could be to come up with what we think is a sensible set of defaults that most folks will use, and then de-prioritise the display settings in the sidebar by having them collapsed by default.

One potential way to handle this would be for us to switch the Display settings panel over to use the new [ToolsPanel component](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-toolspanel--default) (props @aaronrobertshaw), so we could expose just a couple of settings, and hide by default "Nesting level" or other advanced controls that we want to de-prioritise. Here's a quick example of how it might look (just hacked together at the moment, but if the panel is the right tool, we could tweak the wording):

| Hide nesting level by default | Nesting level option available in the ToolsPanel menu |
| --- | --- |
| <img src="https://user-images.githubusercontent.com/14988353/137649745-0ab914e5-6123-4de6-8921-94406bfcb04c.png" width="300" /> | <img src="https://user-images.githubusercontent.com/14988353/137649756-fa75f4be-e650-4a17-bd04-278c6e13a677.png" width="300" /> |

If that approach seems viable, we'd then come up with the list of settings we want to expose by default, and then keep the other settings hidden behind the menu.

karmatosed commented:
> One potential way to handle this would be for us to switch the Display settings panel over to use the new ToolsPanel component (props @aaronrobertshaw), so we could expose just a couple of settings, and hide by default "Nesting level" or other advanced controls that we want to de-prioritise. Here's a quick example of how it might look (just hacked together at the moment, but if the panel is the right tool, we could tweak the wording)

Whilst that is easier than seeing it all, I still think we can remove some settings to get back to a MVP. For example:

- Separator option: I don't think by default that might be needed. Developers could always extend, block styles can be created.
- Show leading separator: I say pick to show or not and then again block style or allow it to just be something that could be adapted.

There are likely more options that can be distilled down, then it's a really easy to use and good block for most cases.

andrewserong commented:
Thanks for the additional feedback, Tammie! There's a bit more to explore here, and with the 5.9 feature freeze in the next week or so, I think it's a good time to park this PR for the moment and then pick it up again in early November after the freeze. It'd be great for us to roll out an MVP in the Gutenberg plugin for a while before we hit a major core release, so that after merge, we can then get some more feedback from folks using the block, which should help us learn more about what's needed before exposing the block more widely.

I'll pick this PR back up again in a couple of weeks, but until then, if anyone has any other feedback or ideas, feel free to drop them here, and we can get back to them.

andrewserong committed:
Remove nestingLevel control

andrewserong committed:
Remove background color control

andrewserong commented:
Hi folks 👋

It's been a few months, and I thought I'd start to dust off this PR again in my spare time. Since it needs a bit of tidying up and fixing before it's ready for review again, I've switched it over to draft. As a quick recap, the next things I'll look into here will be to simplify and stabilise how the block works:

* [x] Remove some unnecessary block supports that are already provided by Group (e.g. border and background color)
* [x] Remove the `nestingLevel` control
* [x] Move separator control to editor canvas via `RichText` control
* [x] Move site title override control to editor canvas via `RichText` control
* [ ] Update tests and fixtures
* [x] Content justification has changed since I last looked at this block, maybe Layout support could work here? (Update: keeping the content justification for the moment, as the flex layout support is more designed for blocks that have inner blocks)
* [x] Update the markup to match the WAI-ARIA [example for breadcrumbs](https://www.w3.org/TR/wai-aria-practices-1.1/#breadcrumb)
* [ ] Re-investigate the block gap support now that it's changed considerably in 5.9
* [ ] Re-do the PR description with fresh overview, details, and screenshots with the simplified approach

I'll ping folks when it's ready for another review! 🙇

talldan commented:
I had a quick look at this yesterday, really impressive work here. This block seems to ship with a lot out of the box, simplifying some aspects like you mention sounds like a good idea and should help make reviews and shipping it easier.

This is not something to worry about for this PR, but something that came to mind is the proposal in https://github.com/WordPress/gutenberg/issues/36667, to make navigation menus one of the primary ways that users shape their site structure in the site editor.

In the future I could imagine breadcrumbs that match a chosen navigation menu's structure being a really useful feature. To do that there probably need to be some better APIs for querying a menu structure (similar to that described in https://github.com/WordPress/gutenberg/issues/30674) so that the link hierarchy can be easily used by this block.

andrewserong committed:
Move separator and site title override controls to the editor canvas via editable RichText fields

andrewserong committed:
Update markup to better match W3C example, try removing spacing support

andrewserong committed:
Allow breadcrumb without parents to render if site title and current page options are enabled

andrewserong committed:
Update test fixtures

paaljoachim commented:
Hey @andrewserong 
I am looking forward to the continuation here!
Getting a good first version in place which we over time can also improve upon.
Should it be brought out of Draft and into a regular PR?

masteradhoc commented:
@talldan @paaljoachim @andrewserong any update on this draft PR? Where are we at?
Would love to see this in core

andrewserong committed:
Fixes after rebase

andrewserong committed:
Fix linting issues

andrewserong commented:
Thanks for the ping @masteradhoc! Unfortunately I haven't had much time to dust off this PR over the past few WP releases as there have been higher priority features and fixes to work on. I'm not sure how much time I'll have to continue working on this PR, but I've at least given this a rebase and fixed up a few linting issues, and I'll switch this over to ready for review.

It could probably use a fresh pair of eyes to test out the current state of the feature, and whether it's a good enough shape for a v1 for core. I'm aware there are lots of plugins that implement Breadcrumbs features, so I'd be curious to hear what folks think might make for a good core version that balances most-needed features without being too opinionated. For example, the current PR adds schema.org markup to the output so that it (hopefully) plays nicely with search engine results, but I'm not too sure if it's a good inclusion for a v1.

aristath commented:
I just did a quick scan of the code so it can't count as a "review", but I do have a couple of notes:

* I'd like to see a filter, to allow changing the `$breadcrumbs` array. This way plugins like WooCommerce and others can just filter the array of breadcrumbs so that the path is OK for all scenarios. This would be a good way to future-proof the implementation, and keep the Core block minimal.
* Ideally we wouldn't want to hardcode the `itemtype="https://schema.org/BreadcrumbList"` in the list. This is a really old method of declaring schema props and in the SEO industry, is sort of considered deprecated. Most SEO plugins generate `JSON-LD` data, so with this block they'll be able to hook to `render_block`, check if the block is a breadcrumb, then add a hook in the `$breadcrumbs` array and generate the proper JSON-LD for the frontend. If however, we have the schema itemtype hardcoded, then we run the risk of search engines detecting duplicate breadcrumbs etc. It would ultimately cause more harm than good.

paaljoachim commented:
Using gutenberg.run to test the most recent version.
It looks like this:
![Screenshot 2023-06-06 at 16 00 54](https://github.com/WordPress/gutenberg/assets/5323259/a2c0868b-05ca-4dd3-ac69-e6c96041ac29)

It looks like a good first version!
I will ping 
@WordPress/gutenberg-design for a look.

I really look forward to having a Breadcrumbs block available in Gutenberg!

jameskoster commented:
I don't know that I'd make this block available in the post editor, it feels more like a theme/template feature to me. In that respect I'd expect it to not only support pages, but posts and taxonomies too. 

In terms of display options, the following are common requests from WooCommerce users:

* Change the label of the home item (site title makes sense as default, but folks often preferred a simple "Home" or even just a 🏠 icon).
* Customise the separator appearance (unsure if there are guidelines around this – it may work better as a filter).

andrewserong committed:
Remove schema markup, only display in site editor

andrewserong committed:
Add filter for inner breadcrumbs array

andrewserong commented:
> I just did a quick scan of the code so it can't count as a "review", but I do have a couple of notes:

Wonderful, thank you for the feedback @aristath, that's exactly the kind of insight I was after! 🙇 Deferring to plugins for SEO support sounds much better, and good to know that adding in a filter should make this straightforward. Also, thanks for the info about microdata vs `JSON-LD` — I knew the latter was more common, but didn't realise the former was effectively deprecated / an old way of doing things 👍

I've removed the schema.org markup and added a filter for `$breadcrumbs` — let me know if you think we need to pass the filter any more info, e.g. the full block object.

> I don't know that I'd make this block available in the post editor, it feels more like a theme/template feature to me. In that respect I'd expect it to not only support pages, but posts and taxonomies too.

Thanks @jameskoster, yes, I was imagining it only available in the site editor — I've pushed an update for that. The current feature set will support both posts and pages. It checks if we're in a hierarchical post type (e.g. pages) and uses that hierarchy, and if not, falls back to using hierarchical categories as in posts. I could imagine potentially adding in custom taxonomies in a follow-up for other use cases?

> In terms of display options, the following are common requests from WooCommerce users:
> * Change the label of the home item (site title makes sense as default, but folks often preferred a simple "Home").
> * Customise the separator appearance (unsure if there are rules around this – it may work better as a filter).

Good to know! In the current state of this PR folks can edit both of these by clicking into them within the editor canvas (for the separator character that means typing in a different character for it), so good to know we've (at least partially) covered those bases.

jasmussen commented:
Nice work. One important aspect of any new block we introduce is to boost the confidence that we actually want it and need it, because they are easy to add, hard to remove.

masteradhoc commented:
@andrewserong 
Thanks for the work and updates! Looks very good! i think its important to also pass by default the breadcrumb schema. ofc we should allow external plugins to modify but this is definately something that will help a lot of sites just by default and there shouldnt be a need of a seperate plugin or modifications to pass a schema in my opinion :)

andrewserong commented:
> One important aspect of any new block we introduce is to boost the confidence that we actually want it and need it, because they are easy to add, hard to remove.

Very good point! Happy to hear feedback from folks as to whether or not, on balance, this is appropriate as a new core block.

>  i think its important to also pass by default the breadcrumb schema.

Thanks for the feedback! I think @aristath outlined well some of the challenges with including the breadcrumb schema in the core block. We currently don't appear to do this for any other core blocks, and to Joen's point about things being easy to add but hard to remove, it could be difficult to remove schema markup after the fact if we wanted to remove it, but would be much easier to add it in a follow up if it turns out that we do need it. So, I think I probably lean toward not including the schema markup for a v1 of the block. It'd then be possible to open follow-ups that explore potentially adding in schema markup (or JSON) to the output, if need be.

I think part of the challenge of this PR in general has been to come up with a good set of defaults that makes for a simple, usable and useful block for most cases, while being flexible enough to be extended and improved over time. Again, please do share any and all feedback! I'll chip away at any feedback and ideas as I have time 🙂

QROkes commented:
It makes no sense to have breadcrumbs without the proper markup.
By the way, schema markup (microdata) is old but not deprecated!

Reference: https://developers.google.com/search/docs/appearance/structured-data/breadcrumb

SEO plugins have their own Breadcrumbs block, so it's important this core block has the minimal markup required. They will not filter this block to add it!

Just for reference, here is what SEO plugins are doing:

- https://yoast.com/features/breadcrumbs/
- https://www.seopress.org/support/guides/enable-breadcrumbs/
- https://rankmath.com/kb/breadcrumbs/

paaljoachim commented:
Hey @andrewserong 
I look forward to the continuation of the work happening with the Breadcrumbs block.
Thanks!

andrewserong commented:
Thanks for the ping, Paal!

Just an update: unfortunately I don't currently have time to continue working on the Breadcrumbs block PR for now, as I've switched to working on other priorities that have design consensus. Before picking this PR back up again, we probably need a decision to be made on the desired scope from a design angle / what folks can agree would be a good MVP for the block in the plugin.

If / when that decision is made, it might be worth dusting this PR off again after WP 6.4 and seeing if it could be made viable again.

Thanks for everyone's continued enthusiasm for the feature! 🙂

paaljoachim commented:
@jameskoster @jasmussen @richtabor 
Now that WP 6.4 has been released. Let's look at the first version of the Breadcrumbs block and get it out into the wild..:)
Thanks!

jarekmorawski commented:
Interesting discussion! There's a clear interest in a Breadcrumbs block. I'm not sure it should be part of the core suite, but there's a chance it could be added when someone browses patterns or blocks in the Inserter (related: https://github.com/WordPress/gutenberg/pull/61504).

After some internal discussion, I put together a design that I hope balances customizability and simplicity. I imagine it'd be extensible, meaning third parties could add custom settings and options (like icons or a new toggle to display the home link as an icon).

![Block](https://github.com/WordPress/gutenberg/assets/79307566/cdbfc394-fcb6-45dc-a7ac-98641a1cdc6c)

The block would consist of links to pages, navigation levels, taxonomies, etc. Using the toolbar, users could set the justification, alignment, and width—same as with the Group block and its variants.

![Layout](https://github.com/WordPress/gutenberg/assets/79307566/4b56ebf5-b0a5-42ae-93e2-e3d352c941d4)

The same settings would also be available in the Inspector. There'd be a new option for `Width` to allow users to enter a custom width in pixels or other units.

![Width](https://github.com/WordPress/gutenberg/assets/79307566/49c74248-a40c-4a54-bde1-f06a3930c22d)

Users would go to Inspector to customize the block's appearance and behavior. When they enable "Show home link", the block will display a link to the home page and users will see an input field in the Inspector. It will allow them to customize the name of the home link. It'd default to the site name but users could enter anything, e.g., `Home` or `Store`.

![Home](https://github.com/WordPress/gutenberg/assets/79307566/7ce565d3-0c3b-4f64-9731-8b254a35b025)

Users could click the ellipsis menu to access more settings, like the current page toggle and a navigation level slider.

![More](https://github.com/WordPress/gutenberg/assets/79307566/b15021fb-ee62-49be-aeb9-6c17fa0f71d4)

When the current page toggle is on, the block will display the name of the currently viewed page as the last breadcrumb level. It will not be active, so it wouldn't have the underline like other breadcrumbs. Users could change its appearance under `Color` in the Style tab.

![Current page](https://github.com/WordPress/gutenberg/assets/79307566/35c72fad-7326-4cc0-9a1f-9e4bb99799be)

The `Max. navigation level `slider would let users decide when breadcrumbs should be truncated. If moved to the left, the block would display `...` in place of the truncated links. Note that the slider would define the _maximum_ number of levels to show, so if there are fewer links, we'd show them all.

![Max levels](https://github.com/WordPress/gutenberg/assets/79307566/000d5477-e851-4f2d-9794-85916d347987)

Users could change the separator in the Styles tab. Of all the dozens of websites I checked, all used one of the three options I included in the design. It could be a good starting point and if users ever need something else, they'd install a plugin that extends the block.

I also included a toggle for displaying the leading separator and reworded the copy a bit, though I'd argue the pattern isn't common enough to be part of the block's default setup. Happy to keep it in the design if you feel it's valuable.

![Separator](https://github.com/WordPress/gutenberg/assets/79307566/db5549b4-4e82-4efc-a1be-942b0e900c2d)

Once again, users could change the separator color in the Color section.

![Color](https://github.com/WordPress/gutenberg/assets/79307566/8f7b55d0-3a73-498d-9c49-0941d7e00197)

Let me know if you have any thoughts!

andrewserong commented:
Thanks again for all the discussion, feedback, and ideas on this PR, everyone! As it's quite old and I see there's renewed interest in the idea of adding in a breadcrumbs block to WordPress, I'll close out this old PR to make it clear that it's no longer current. Let's continue on discussing this block over in the linked issue (https://github.com/WordPress/gutenberg/issues/21943).
